src/db/file.rs

```diff
@@ -138,102 +141,96 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
-                        body: Some(content.clone().into()),
-                        content_type: Some(mime.clone()),
-                        ..Default::default()
-                    }).sync();
-                    attempts += 1;
-                    match s3_res {
-                        // we've successfully uploaded the content, so steal it;
-                        // we don't want to put it in the DB
-                        Ok(_) => break None,
-                        // Since s3 was configured, we want to panic on failure to upload.
-                        Err(e) => {
-                            log::error!("failed to upload to {}: {:?}", bucket_path, e);
-                            // Get a new client, in case the old one's connection is stale.
-                            // AWS will kill our connection if it's alive for too long; this avoids
-                            // that preventing us from building the crate entirely.
-                            *client = s3_client().unwrap();
-                            if attempts > 3 {
-                                panic!("failed to upload 3 times, exiting");
-                            } else {
-                                continue;
-                            }
-                        },
-                    }
-                }
             } else {
-                Some(content.clone().into())
-            };
-
-            file_list_with_mimes.push((mime.clone(), file_path.clone()));
-
-            (
-                bucket_path,
-                content,
-                mime,
-            )
+                mime.to_owned()
+            }
         };
 
-        // If AWS credentials are configured, don't insert/update the database
-        if client.is_none() {
+        if let Some(client) = s3_client() {
+            let bucket_path = bucket_path.clone();
+            let content = content.clone();
+            let mime = mime.clone();
+
+            futures.push(client.put_object(PutObjectRequest {
+                bucket: "rust-docs-rs".into(),
+                key: bucket_path.clone(),
+                body: Some(content.clone().into()),
+                content_type: Some(mime.clone()),
+                ..Default::default()
+            }).map_err(move |e| {
+                log::error!("failed to upload to {}: {:?}", bucket_path, e);
+                client.put_object(PutObjectRequest {
+                    bucket: "rust-docs-rs".into(),
+                    key: bucket_path,
+                    body: Some(content.into()),
+                    content_type: Some(mime),
+                    ..Default::default()
+                }).map_err(|_| {
+                    panic!("failed to upload twice, exiting");
+                })
+            }));
+
+            if futures.len() > MAX_CONCURRENT_UPLOADS {
+                if rt.block_on(::futures::future::join_all(futures)).is_err() {
+                    panic!("upload to s3 failed");
```

Comment:
Should we include the error details in this panic message? I feel like that would make diagnosing issues much easier.

Reply:
what details would you like to see added?

Reply:
debug print of the error is probably enough -- instead of is_err we can do if let Err(e) = ... { panic!("upload to s3 failed: {:?}", e); } or so

Reply:
didn't mean to start a review...

```
help: the trait `std::fmt::Debug` is not implemented for `rusoto_core::RusotoFuture<rusoto_s3::PutObjectOutput, rusoto_s3::PutObjectError>`
```

I think if I can unwrap that `RusotoFuture` to get the `PutObjectError` I will be able to accomplish this. I'm not sure how to do that exactly.

Reply:
That seems a bit odd -- block_on shouldn't be returning a future I'd imagine? Maybe your parentheses are off or something?

Reply:
It might be a result:

https://github.com/rusoto/rusoto/blob/df96e8ce92f34248ec667f8514817158bb06ff01/rusoto/core/src/future.rs#L275

Reply:
Well in any case -- it sounds like this might be harder to do that @QuietMisdreavus expected so probably not worth it then. Most of the time AWS errors are really opaque anyway...

Reply:
Hmm, now that i look at the future being inserted into the queue, it's already panicking if it's failing twice:

```rust
            futures.push(client.put_object(PutObjectRequest {
                bucket: "rust-docs-rs".into(),
                key: bucket_path.clone(),
                body: Some(content.clone().into()),
                content_type: Some(mime.clone()),
                ..Default::default()
            }).map_err(move |e| {
                log::error!("failed to upload to {}: {:?}", bucket_path, e); // logging the error after the first failure
                client.put_object(PutObjectRequest {
                    bucket: "rust-docs-rs".into(),
                    key: bucket_path,
                    body: Some(content.into()),
                    content_type: Some(mime),
                    ..Default::default()
                }).map_err(|_| {
                    panic!("failed to upload twice, exiting"); // already panicking after the second error
                })
            }));
```

I wonder if the nesting should be changed on that. As far as i remember combinator chaining working out, you can tag on the second `map_err` without introducing that rightward-drift:

```rust
            futures.push(client.put_object(PutObjectRequest {
                bucket: "rust-docs-rs".into(),
                key: bucket_path.clone(),
                body: Some(content.clone().into()),
                content_type: Some(mime.clone()),
                ..Default::default()
            }).map_err(move |e| {
                log::error!("failed to upload to {}: {:?}", bucket_path, e);
                client.put_object(PutObjectRequest {
                    bucket: "rust-docs-rs".into(),
                    key: bucket_path,
                    body: Some(content.into()),
                    content_type: Some(mime),
                    ..Default::default()
                })
            }).map_err(|e| {
                log::error!("failed to upload to {} a second time: {:?}", bucket_path, e);
                panic!("failed to upload twice, exiting");
            }));
```

Would something like this work?

Reply:
I'll give that a shot

Reply:
Nope, same issue...

```
help: the trait `std::fmt::Debug` is not implemented for `rusoto_core::RusotoFuture<rusoto_s3::PutObjectOutput, rusoto_s3::PutObjectError>`
```